tests/docker/Dockerfile

```diff
@@ -0,0 +1,98 @@
+#
+# This file is part of mbed TLS (https://tls.mbed.org)
+#
+# Copyright (c) 2018, ARM Limited, All Rights Reserved
+#
+# Purpose
+#
+# Defines a Docker container suitable to build and run all tests (all.sh),
+# except for those that use a proprietary toolchain.
+
+FROM ubuntu:xenial
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN apt update \
+    && apt -y install software-properties-common \
+    && rm -rf /var/lib/apt/lists
+
+RUN add-apt-repository -y ppa:team-gcc-arm-embedded/ppa
+
+RUN apt update \
+    && apt -y install \
+    # mbedtls build/test dependencies
+    build-essential \
+    clang \
+    cmake \
+    doxygen \
+    gcc-arm-none-eabi \
+    gcc-mingw-w64-i686 \
+    gcc-multilib \
+    g++-multilib \
+    gdb \
+    git \
+    gnutls-bin \
+    graphviz \
+    lsof \
+    python \
+    python-pip \
+    python3 \
+    wget \
+    # libnettle build dependencies
+    libgmp-dev \
+    m4 \
+    pkg-config \
+    && rm -rf /var/lib/apt/lists/*
+
+# Build a static, legacy openssl from sources with sslv3 enabled
+# Based on https://gist.github.com/bmaupin/8caca3a1e8c3c5686141 (build-openssl.sh)
+RUN cd /tmp \
+    && wget https://openssl.org/source/openssl-1.0.2k.tar.gz -qO- | tar xz \
+    && cd openssl-1.0.2k \
+    && ./config --prefix=`pwd`/local --openssldir=/usr/lib/ssl enable-ssl2 enable-ssl3 enable-weak-ssl-ciphers no-shared \
+    && make depend \
+    && make \
+    && make -i install_sw \
+    && cp local/bin/openssl /usr/local/bin/legacy-openssl \
+    && rm -rf /tmp/openssl*
+
+# Build libnettle 2.7.1 from sources; legacy gnutls needs it
+RUN cd /tmp \
+    && wget https://ftp.gnu.org/gnu/nettle/nettle-2.7.1.tar.gz -qO- | tar xz \
+    && cd nettle-2.7.1 \
+    && ./configure --disable-documentation \
+    && make -j \
```

Comment:
`all.sh` also applies `-j` unless overridden, and `-j` makes for a significant speedup here. Timings on my PC:

Step | With -j | Without
----- | --- | ---
nettle | 12.4s | 43.7s
gnutls-3.3.18 | 46.9s | 78.7s
openssl-1.1.1a | 36.3s | 164.7s

The CPU usage is somewhat bursty so it _should_ cooperate decently enough with other processes. But I can expose a build arg into docker_env.sh to support overrides. Looking more closely it should probably be `-j $(nproc)` anyway.

As to broken makefiles- I've run these make commands a couple dozen times in the course of developing this PR. The only one that has failed for me is openssl-1.0.2g and I've special-cased it.

That has me thinking though- MAKEFLAGS should propagate into the `docker container run` as well to have the expected effect.

Reply:
Thanks for the additional analysis! Regarding the makeflags - if you mean passing `MAKEFLAGS=-jX`, or even `MAKEFLAGS=-j$(nproc)` to the docker build command, and then override the value for openssl 1.0.2, and whatever doesn't build in parallel, then I think it's a good idea (especially, that in your setup, you can conveniently do this from the bash scripts). However, since, as you have noticed, it may break the builds in some cases, I think it would be the safest path, to have a CLI switch to disable all the `make` parallelization at the scripts level. It should also be configurable (on/off) in case somebody wants to use the plain dockerfile (i.e. it shouldn't be only possible via the bash scripts).

Reply:
Parallelism of make during the `docker image build`  is now controlled with the `--build-arg MAKEFLAGS_PARALLEL` (in `docker_env.sh`). docker_env will also pass MAKEFLAGS to the `docker container run`, if it is set.